### PR TITLE
Allow dictionary of per-term typifical functions

### PR DESCRIPTION
**Pull request by palday**

### CODE REVIEW

File: src/regressionmodel.jl [hunk 1]
```diff
@@ -12,6 +12,15 @@ is specified via `typical`. Note that this is also applied to categorical contra
...
 set used to fit the model.
 
+`typical` can either be a single, scalar-valued function or a dictionary
+matching term symbols to scalar-valued functions. The use of a dictionary
+allows specifying different `typical` functions for different input variables.
+In this case, `typical` functions must be provided for all term variables
+except the intercept. If there is a single term that should be "typified"
+differently than others, then the use of `DataStructures.DefaultDict` may
+be useful to create a default `typical` value with only the exception
+explicitly specified.
```
Comment (kolia) [inline src/regressionmodel.jl:22]:
```suggestion
explicitly specified. For example:

```julia
using DataStructures
typical = DefaultDict(() -> mean)  # default to x -> mean(x)
typical[:sex] = v -> 0.0                   # typical value for :sex
```
```

(Not sure how to escape triple quotes inside a github suggestion, the closing quotes get dropped...)

Comment (kleinschmidt) [reply src/regressionmodel.jl:22]: just use more quotes ;) 

`````

````
```julia
using Stuff
```
````

`````

Comment (kleinschmidt) [reply src/regressionmodel.jl:22]: (that's quintuple backquotes around the whole thing)

File: src/regressionmodel.jl [hunk 2]
```diff
@@ -12,6 +12,21 @@ is specified via `typical`. Note that this is also applied to categorical contra
 thus yielding an average of the contrast, weighted by the balance of levels in the data
 set used to fit the model.
 
+`typical` can either be a single, scalar-valued function or a dictionary
```
Comment (kleinschmidt) [inline src/regressionmodel.jl:15]:
```suggestion
`typical` can either be a single, scalar-valued function (e.g. `mean`) or a dictionary
```

File: src/typical.jl [hunk 3]
```diff
@@ -103,7 +103,17 @@ function _trmequal(t1::FunctionTerm, t2::FunctionTerm)
...
 end
 
-function typicalterm(term::AbstractTerm, context::MatrixTerm, model_matrix; typical=mean)
+# XXX should probably drop the kwarg version of this
+# and skip the redirect, but that's Technically Breaking™
+# likewise add an AbstractMatrix type restriction to model_matrx
+function typicalterm(term::AbstractTerm, context::MatrixTerm,
+                     model_matrix; typical=mean)
+    return typicalterm(term, context, model_matrix, typical)
+end
```
Comment (kleinschmidt) [inline src/typical.jl:112]: use `@deprecate` then?

```suggestion
@deprecate typicalterm(term::AbstractTerm, context::MatrixTerm, model_matrix; typical=mean) typicalterm(term, context, model_matrix, typical)
```

dunno if this actually works as written but that's the idea...

**Review by kleinschmidt — APPROVED** [review]
this looks good.  one tiny suggestion but this seems like a reasonable approach.

File: src/typical.jl [hunk 4]
```diff
@@ -103,7 +103,11 @@ function _trmequal(t1::FunctionTerm, t2::FunctionTerm)
            t1.forig == t2.forig
 end
 
-function typicalterm(term::AbstractTerm, context::MatrixTerm, model_matrix; typical=mean)
+@deprecate typicalterm(term::AbstractTerm, context::MatrixTerm, model_matrix; typical=mean) typicalterm(term, context, model_matrix, typical)
```
[the remark this answers is not in the document]
Comment (kleinschmidt) [reply src/typical.jl]: AHHHHHHHH my eyes

**Pull request merged**